### ISSUE TITLE
Implement way to cherry pick usptream commits in ODH

### ISSUE THREAD

**Issue by Gregory-Pereira** [open-services-group/devsecops#11]
We want to be able to integrate upstream changes in ODH-Manifests into our deployed version. However we want to keep the commit SHAs the same as in upstream to verify the integrity of our code base, and to make it easier to compare against upstream. Our current plan for this is to update the [AICoE z-stream RHODS task](https://github.com/AICoE/aicoe-ci/blob/master/tasks/z-stream-update-rhods-task.yaml) to be able cherry pick commits.

To that end, we need to test the AICoE task and make any necessary changes to move it to a production setting.

/assign

**Comment by Gregory-Pereira** [comment]
More information available [here](https://github.com/open-services-group/devsecops/issues/9#issuecomment-1063159124)

**Comment by Gregory-Pereira** [comment]
When attempting to implement this, I came across an issue with the git histories that I will attempt to describe here.

# Scenario 

So lets say we wanted to create a new branch, which would be `$cluster-$odh_version`. Some examples of this might be `smaug-v1.1.0` , `osc-cl1-v1.1.2`, or any of our other existing tags: [`osc-cl1-v1.1.0`, `osc-cl2-v1.1.0`, `osc-cl2-v1.1.2`]. So as I see it there are two different cases to consider:

1. Deployment of ODH to a cluster that has no previous deployment of ODH with a version of ODH that we have previously used as the base of one of our other branches (ex: `smaug-v1.1.0` because `v1.1.0` was used in `osc-cl1-v1.1.0` and `osc-cl2-v1.1.0`).
2. Upgrading our deployment of ODH that already exists on a given cluster based on a new tagged version of ODH (ex: `osc-cl2-v1.1.2` because `osc-cl2-v1.1.0` already exists.)

Before I step through the cases remember that our original process for creating a new branch would be as follows.
1. create branch from `opendatahub-io/odh-manifests` tag
2. cherry pick commits that are related to operate-first deployment and not `opendatahub-io/odh-manifests` onto that branch
3. push branch to `operate-first/odh-manifests`.

##  Finding the problem

Now the process for both cases would be pretty similar because in both cases, a branch in `operate-first/odh-manifests` must already be created with said tagged version of ODH (ex: to make branch `smaug-v1.1.0`, `osc-cl1-v1.1.0` already exists, or to make branch `osc-cl1-v1.1.2`, branch `osc-cl2-v1.1.2` already exists). That being the case, our pipeline / task would assume that there is an "example branch" to look at, where it can pull the operate-first related commit hashes. However, in cherry-picking those commits, whatever runs the task (shesheta, kebechet, any other bot, or a user via a personal access token) would be the one who is actually committing the changes authored by someone else in another commit. For an example of this see [osc-cl1-v1.1.2 branch moved manifests commit](https://github.com/operate-first/odh-manifests/commit/6cf1118bfa7e18bb557483e494459d13efc6fca0) authored by Gage, where Humair is considered the commiter.

Commit SHAs kinda work like a block in the blockchain, and each hash is computed from hashing the 5 following catagories or peices of information:
    1. The previous git history / tree
    2. Parent commit sha
    3. Commit message
    4. Commit author details
    5. Commiter details

## The Problem

Therefore while 1-4 would be the same if one of us introduced commits in a branch of `operate-first/odh-manifests` verses if this task ran in Tekton, and an automated tool cherry-picked and committed those changes onto `operate-first/odh-manifests`, the committer details will never be the same (unless we used humairs PAT to make all the commits). This means that the commit SHA / history will be the same up untill we introduce the commits that reconcile the difference between operate first's and opendatahub's deployments of ODH, but will have comepletely different histories after this, even for different cluster's versions of the same code / odh-version.

TLDR: we will never really be able to maintain the history and not adjust the commit SHAs unless one person would always be the one to commit the changes that reconcile the difference in operate-first's and opendatahub's ODH deployment, and we were to use their PAT for the tekton task to do the cherry picking. Even then, this would introduce other issues, like how the bot would be pushing code changes as it's own, and potentially attribution (if we didn't opt for the persons PAT).

## Rethinking the issue

Firstly, I think we should end up dropping the whole cherry-pick idea and instead opt for just rebasing. Its much easier, and what is the point of going to the extra effort in cherry-picking when it won't even maintain the SHA's of the commits we are adding in. 

Secondly, we also shouldn't start with an tagged version of ODH as the base of our branch, instead we should base our branch off of another "example branch" with the changes for the same version of odh that you are seeking to deploy. Then we should find the differences in commits between the verison of ODH that the previous branch pointed to (ex: `osc-cl1-1.1.0` --> ODH v1.1.0) and the desired version of ODH on the new branch (ex: ODH v1.1.2) , and rebase those onto our our coppied new branch (ex: `osc-cl1-v1.1.2` which was created from `osc-cl1-v1.1.0`).

I also envision a new use case for this task. It would be very similar to use case 2 mentioned above, but for a specific component:

> Upgrading our deployment of ODH that already exists on a given cluster based on a new tagged version of ODH (ex: `osc-cl2-v1.1.2` because `osc-cl2-v1.1.0` already exists.)


Instead this new case would do this, but would only cherry-pick / rebase the commits that modified files to an specific component of ODH (ex: jupyterhub). This is different from use cases 1 and 2, because the result of those cases would be a PR generated by the bot with all the commits that differ between the two versions of ODH in upstream (ex: diff in commits of  `v1.1.0` and `v1.1.2`). It would then be up to the maintainers / operators of `operate-first/odh-manifests` to go through the PR generated by the bot, and select the commits depending on if they are related to the components we are currently deploying on the that cluster's version of ODH. Instead, this 3rd case would need no editing and could be made entirely by the bot, because you would specify a component or components, and only the commits that modify that/those components would be included in the PR.

Would love to hear some others thoughts on this. I will move forward this by attempting to implement it in both ways, so that while we reach a consensus on `rebase` vs. `cherry-pick`, we can have something we can test that hopefully works.

**Comment by harshad16** [comment]
This discussion would be carried on the operate-first/odh-manifests repository  as per requirements.
This not critical as of the moments
Thanks for the effective discussion :100:
